### ISSUE TITLE
DateTime.AddMonths calculation is incorrect

### ISSUE THREAD

**Issue by JechoJekov** [Saltarelle/SaltarelleCompiler#253]
`DateTime.AddMonths` performs the calculations by using the local time. However, this does not takes into an account daily saving time transition.

For example, the following code:

```
var time = DateTime.FromUtc(2013, 6, 1);
var time2 = time.AddMonths(6);
System.Console.WriteLine(time.ToString() + " | " + time.ToUtcString());
System.Console.WriteLine(time2.ToString() + " | " + time2.ToUtcString());
```

Outputs

```
Sat Jun 01 2013 03:00:00 GMT+0300 (FLE Daylight Time) | Sat, 01 Jun 2013 00:00:00 GMT
Sun Dec 01 2013 03:00:00 GMT+0200 (FLE Standard Time) | Sun, 01 Dec 2013 01:00:00 GMT 
```

while the correct result (second line) should be:

```
Sun Dec 01 2013 02:00:00 GMT+0200 (FLE Standard Time) | Sun, 01 Dec 2013 00:00:00 GMT 
```

The reason for this behavior is the `DateTime.AddMonths` uses local time components instead of UTC ones. More specifically the code is the following:

```
[InlineCode("new Date({this}.getFullYear(), {this}.getMonth() + {months}, {this}.getDate(), {this}.getHours(), {this}.getMinutes(), {this}.getSeconds(), {this}.getMilliseconds())")]
public DateTime AddMonths(int months)
{
    return default(DateTime);
}
```

The correct code is:

```
[InlineCode("new Date(Date.UTC({this}.getUTCFullYear(), {this}.getUTCMonth() + {months}, {this}.getUTCDate(), {this}.getUTCHours(), {this}.getUTCMinutes(), {this}.getUTCSeconds(), {this}.getUTCMilliseconds()))")]
public DateTime AddMonths(int months)
{
    return default(DateTime);
}
```

Since the month during which the transition to/from DST is performed may change over the years or DST transition may not be performed at all during some years (in some parts of the world), the `DateTime.AddYears` should be corrected as well:

```
[InlineCode("new Date(Date.UTC({this}.getUTCFullYear() + {years}, {this}.getUTCMonth(), {this}.getUTCDate(), {this}.getUTCHours(), {this}.getUTCMinutes(), {this}.getUTCSeconds(), {this}.getUTCMilliseconds()))")]
public DateTime AddYears(int years)
{
    return default(DateTime);
}
```

Regards,
Jecho

**Comment by JechoJekov** [comment]
Merge request created.

**Comment by JechoJekov** [comment]
Can you delete commit "1eda354f3a84a6e92856444e8060e29355b0948c"? The code is not correct.

**Comment by JechoJekov** [comment]
I have developed a class called "DateTimeExt" whose behavior is closer to DateTime in .NET Framework. It uses "DateTimeKind" similar to .NET in order to provide better support for UTC time.
It is implemented in C#, however, if there is a consensus to include it in the runtime library I can translate it to JS.

**Comment by erik-kallen** [comment]
Kind of depends on how it looks, IMO. Is it online?

**Comment by JechoJekov** [comment]
Posted here: http://collabedit.com/su22b

The basic idea is that Date in JS is already in UTC time so depending on the DateTimeKind value the property should return either local or UTC value.

**Comment by erik-kallen** [comment]
Perhaps we could do something quirky like adding an optional property 'kind' to normal DateTimes (eg. ```[InlineCode("{this}.kind || Unspecified")] public DateTimeKind Kind { get; set; }```)

**Comment by JechoJekov** [comment]
Yes - that would be the best approach. However, the properties "Year", "Month", "Hour", etc. should be changed to return correct values. Basically there are two views of a "DateTime" instance - local and UTC depend on the value of the kind field.

**Comment by erik-kallen** [comment]
I agree. DateTime should work as the .net type; people who want JS behavior can use JsDate.

**Comment by JechoJekov** [comment]
OK. Then I'll prepare pull requests with changes to DateTime and the runtime library. Although removing methods not available in .NET would result in better design keeping them would provide compatibility with previous versions.

**Comment by JechoJekov** [comment]
I'd personally would vote to remove methods not available in .NET, but obviously it is your call.

**Comment by erik-kallen** [comment]
Which methods are you talking about?

Perhaps we should do this in v3 only.

**Comment by JechoJekov** [comment]
I'm referring to "GetXXX", "GetUtcXXX", "FromUtc", "ParseExactUtc" and any other methods that are not available in .NET Framework.

**Comment by erik-kallen** [comment]
Yes, we are probably best off removing those if we are introducing the Kind property.

**Comment by JechoJekov** [comment]
OK. Then shall I create a pull request with changes to DateTime and the runtime library? Or perhaps introduce it as a separate class until v3 is ready?

**Comment by erik-kallen** [comment]
Can you create a PR to the Roslyn branch?

**Comment by JechoJekov** [comment]
OK
